### ISSUE TITLE
Propose to officially support ARM64 platform

### ISSUE THREAD

**Issue by liusheng** [grpc/grpc-java#6364]
### Is your feature request related to a problem?

Yes, recently, I am working on promoting Hadoop building on ARM64 server, while an error occured due to missing the `io.grpc:protoc-gen-grpc-java:exe:linux-aarch_64:1.15.1` artifact for aarch platform, see error messages:
```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  03:13 min
[INFO] Finished at: 2019-05-17T09:28:55+08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.xolstice.maven.plugins:protobuf-maven-plugin:0.5.1:compile-custom (default) on project hadoop-yarn-csi: Missing:
[ERROR] ----------
[ERROR] 1) io.grpc:protoc-gen-grpc-java:exe:linux-aarch_64:1.15.1
[ERROR]
[ERROR]   Try downloading the file manually from the project website.
[ERROR]
[ERROR]   Then, install it using the command:
[ERROR]       mvn install:install-file -DgroupId=io.grpc -DartifactId=protoc-gen-grpc-java -Dversion=1.15.1 -Dclassifier=linux-aarch_64 -Dpackaging=exe -Dfile=/path/to/file
[ERROR]
[ERROR]   Alternatively, if you host your own repository you can deploy the file there:
[ERROR]       mvn deploy:deploy-file -DgroupId=io.grpc -DartifactId=protoc-gen-grpc-java -Dversion=1.15.1 -Dclassifier=linux-aarch_64 -Dpackaging=exe -Dfile=/path/to/file -Durl=[url] -DrepositoryId=[id]
[ERROR]
[ERROR]   Path to dependency:
[ERROR]         1) org.apache.hadoop:hadoop-yarn-csi:jar:3.3.0-SNAPSHOT
[ERROR]         2) io.grpc:protoc-gen-grpc-java:exe:linux-aarch_64:1.15.1
[ERROR]
```
In the Maven central repo, there are different version of  `protoc-gen-grpc-java` artifacts for linux-x86, osx-x86, windows-x86, but no `aarch64` specific package. see:
https://search.maven.org/search?q=g:io.grpc%20AND%20a:protoc-gen-grpc-java&core=gav

I have searched this problem in google, found that there are some similar issues brought up by others, such as https://github.com/grpc/grpc-java/issues/2202 and [disscussion](https://groups.google.com/forum/#!topic/grpc-io/lEwfOdDgdZU/discussion) , there are also some posts share workarounds by compiling grpc-java manually with private patches. actually, I have also tried to build `protoc-gen-grpc-java` but it also easily fail for some reason.

### Describe the solution you'd like

Hope we can officially support ARM64 platform and publish `io.grpc:protoc-gen-grpc-java:exe:linux-aarch_64` artifact in Maven central repo.
I am not sure if I am missing someting, please pointing out if I am wrong.

### Describe alternatives you've considered

NA
### Additional context

NA

**Comment by ejona86** [comment]
Even after compiling you will find that running may not work because netty_tcnative (and Conscrypt) isn't pre-built for ARM64. Will this environment be using Java 9+? Java 9+ is required for running on ARM64 today.

Can you explain/link to the benefit of building Hadoop on ARM64? Is this more for servers or hobbyists?

We don't actually have an ARM64 environment at the moment, especially one that would be easy to integrate with our release process. So we might have to use an emulator or cross-compile. So it would take some work.

**Comment by liusheng** [comment]
Hi @ejona86,

Thank you for your attention.

After apply my manually built `protoc-gen-grpc-java`, I can sucessfully build Hadoop and run Hadoop YARN unit tests, but didn't do further validation. As you pointed out, the netty_tcnative(and Conscrypt) may also block, I will check more later and to see if there is possible also to promoting netty_tcative officially suport. I am using a server with Ubuntu 16.04 and using java 8, I am not sure if there is guide show that Java 9+ is required on ARM64, but seems here https://openjdk.java.net/projects/aarch64-port/ indicate that java 8 is also support, I will try again with java 9+ :)

No this is not a hobbyist :), currently, there are growing numbers of ARM based datacenter servers showing up on the market, the performance and stability has been continuously improving, and ARM architecture is the dominate arch for devices at edge, like IOT,  autonomous vehicle, surveillance cameras etc. Enabling big data platform like Hadoop on those devices will enrich the overall pplication scenario, bring possiblities and opportunities for new buissiness models and probably provide solutions for some known bottlenecks.

Actually, I and my colleague are working on promoting some popular opensource projects to adapt ARM64 platform recently, such as Hadoop, Spark, Flink, we have a pool of  ARM resources to support this work. If it is needed here, we are very pleasure to provide an ARM server for grpc-java community so we don't need emulator or cross-compile.

**Comment by ejona86** [comment]
We generally require Java 9+ for running _grpc-java_ on ARM64. Although it's technically only necessary if using TLS (because tcnative doesn't have pre-built binaries). Nitty-gritty is in [SECURITY.md](https://github.com/grpc/grpc-java/blob/master/SECURITY.md#tls-on-non-android). I guess maybe Hadoop uses plaintext commonly, so you might not have noticed.

> especially one that would be easy to integrate with our release process

This was key. I was more concerned with the integration work. It would probably be easiest to cross-compile using gcc in a Docker image than using a new machine.

A new machine would take work to trigger jobs on it during our release/CI process and it would take work to grab the results from the machine to combine them with results from other platforms.

**Comment by liusheng** [comment]
Hi @ejona86,
Sorry for replying lately, thank you for pointing out, I can find there is java 9+ available in [launchpad.net](https://launchpad.net/ubuntu/+source/openjdk-lts/10.0.1+10-3ubuntu1/+build/14783735), but seems not in debian packages repository, I will test later.

For integration work, not sure whether the cross-compile approach can fully do building and validation or not. but if you need, we can long-term provide and maintain a ARM64 VM for grpc community.

**Comment by ejona86** [comment]
If we are targeting Java 11, then the only real thing necessary to build grpc-java-using projects on ARM64 is the protoc plugin. `protoc` itself is already being built for other platforms, so I don't think any other projects would need to add ARM64 for grpc-java.

https://github.com/grpc/grpc-java/blob/v1.25.0/buildscripts/kokoro/linux_artifacts.sh is the script that would need to be updated. We would need to build libprotoc and then the compiler. Protobuf's own build can be a reference: https://github.com/protocolbuffers/protobuf/blob/v3.10.1/kokoro/release/protoc/linux/build.sh . It cross-compiles, which is my strong preference.

**Comment by liusheng** [comment]
Hi @ejona86,

I have tried to build the grpc-java of version `1.15.1`, which version is required by Hadoop, with Java 11 on a Ubuntu 18.04 server, but still met some issues，see my env info:
```
------------------------------------------------------------         
Gradle 4.9                                                           
------------------------------------------------------------         
                                                                     
Build time:   2018-07-16 08:14:03 UTC                                
Revision:     efcf8c1cf533b03c70f394f270f46a174c738efc               
                                                                     
Kotlin DSL:   0.18.4                                                 
Kotlin:       1.2.41                                                 
Groovy:       2.4.12                                                 
Ant:          Apache Ant(TM) version 1.9.11 compiled on March 23 2018
JVM:          11.0.4 (Ubuntu 11.0.4+11-post-Ubuntu-1ubuntu218.04.3)  
OS:           Linux 4.15.0-29-generic aarch64                  
```
---
**Issue 1:**
building doc failure
```
The code being documented uses modules but the packages defined in https://docs.oracle.com/javase/8/docs/api/ are in the unnamed module
```
there is an existing issue: exist similar issue: https://github.com/joel-costigliola/assertj-core/issues/1403
**workaround:** add `-x javadoc` arguments to `gradlew` command to skip doc generation

---
**issue2:**
missing `protobuf-javalite` of aarch64 dependency
```
3. Execution failed for task ':grpc-protobuf-lite:generateTestProto'.                                                                                                      
> Could not resolve all files for configuration ':grpc-protobuf-lite:protobufToolsLocator_javalite'.                                                                    
   > Could not find protoc-gen-javalite-linux-aarch_64.exe (com.google.protobuf:protoc-gen-javalite:3.0.0).                                                            
     Searched in the following locations:                                                                                                  
         https://maven-central.storage-download.googleapis.com/repos/central/data/com/google/protobuf/protoc-gen-javalite/3.0.0/protoc-gen-javalite-3.0.0-linux-aarch_64.exe
```
**workaround:** compile protobuf-v3.0.0-javalite manually and specify in gradle command.

---
**issue3:** 
```
> Task :grpc-alts:compileJava FAILED                                                                                                                        
/home/liusheng/grpc-java/alts/src/generated/main/grpc/io/grpc/alts/internal/HandshakerServiceGrpc.java:20: error: cannot find symbol                         
@javax.annotation.Generated(                                                                                                                                
                 ^                                                                                                                                         
  symbol:   class Generated                                                                                                                                  
  location: package javax.annotation
```
there are some similar issues and discussions in grpc-java community, I have tried some suggestions, but seems cannot work.

Although the building process failed because of the above `issue 3`, but the `protoc-gen-grpc-java` binary has been compiled.
```
liusheng@ecs-791c-0003:~/grpc-java$ ll compiler/build/exe/java_plugin/protoc-gen-grpc-java              
-rwxrwxr-x 1 liusheng liusheng 2362616 Nov 11 19:13 compiler/build/exe/java_plugin/protoc-gen-grpc-java*
```
FYI, [here](https://gist.github.com/liusheng/d6dd509a371a9b7b07b75673646520b2) is my build script. 

Could please give me some suggestions ? how about we publish the `protoc-gen-grpc-java-1.15.1` to maven central repo? that make it out-of-box for Hadoop usage :)

Thanks!

**Comment by ejona86** [comment]
Don't use v1.15.1. We aren't going to make any changes to v1.15 for this. We aren't going to publish new binaries for v1.15. Use the latest release. v1.15 was cut before Java 11 was released, so there will be pain there. https://github.com/grpc/grpc-java/commit/959323be was in v1.17. But really, just use the latest release.

The javalite issue is expected. Yeah, you get to build it yourself which is annoying. We may be able to bump javalite to a newer version soon that has an arm64 build.

Note that you only need to run `./gradlew :grpc-compiler:build`, not the full build.

And again, integrating a new machine into our build process would take some real effort and coordination. The release builds should be cross-compiled.

**Comment by kiwik** [comment]
I am trying to add ARM64 supporting for codegen plugin too. According to document ["Build Artifacts"](https://github.com/grpc/grpc-java/blob/master/RELEASING.md#build-artifacts) , we should modify file `linux_artifacts.sh` and apply cross-compiled in order to integrating with automation process, and I will refer to protoc document to test this way in my development machine first.

**Comment by liusheng** [comment]
Hi @ejona86,
I have added a change about the build scripts to support cross compiling for aarch64 platform. see: #6441, the building process need to build `protobuf` as first step, which using the docker image of `Centos 6.9`, see: https://github.com/protocolbuffers/protobuf/blob/master/protoc-artifacts/Dockerfile but the `Centos 6.9` seems cannot correctly install `g++-aarch64-linux-gnu` which is required for aarch64 cross compiling. even if I installed it, the cross building process of protobuff raised error. So, in this change, I just directly put the building process in building server on [recommend build environment Ubuntu 14.04](https://github.com/grpc/grpc-java/blob/master/RELEASING.md#build-environments) and outside docker. not sure if this is OK ?

Another problem, after apply this change, the building process of protobuf is OK, but the building process of grpc-java will still raise error, see the [error message](http://paste.openstack.org/show/786262/), any suggestion ?

Thanks!

**Comment by liusheng** [comment]
Hi @ejona86,
I have updated the change #6441, now the script can successfully cross compile the `protoc-gen-grpc-java` artifact, but after bulding `protoc-gen-grpc-java` the building process will use the artifact as plugin of `protoc` to generate the tests proto files, because the compiled `protoc-gen-grpc-java` is for `aarch64` platform, so it cannot be executed in x86_64 server.  see the error:
```
> Task :grpc-compiler:generateTestProto FAILED
Caching disabled for task ':grpc-compiler:generateTestProto' because:
  Build cache is disabled
Task ':grpc-compiler:generateTestProto' is not up-to-date because:
  Task has failed previously.
[/root/.gradle/caches/modules-2/files-2.1/com.google.protobuf/protoc/3.10.0/e7ba2967f692c3bd2bd417f6a305bcb6a8c6a357/protoc-3.10.0-linux-x86_64.exe, -I/grpc-java/compiler/src/test/proto, -I/grpc-java/compiler/build/extracted-protos/test, -I/grpc-java/compiler/build/extracted-include-protos/test, --java_out=/grpc-java/compiler/build/generated/source/proto/test/java, --plugin=protoc-gen-grpc=/grpc-java/compiler/build/exe/java_plugin/protoc-gen-grpc-java, --grpc_out=/grpc-java/compiler/build/generated/source/proto/test/grpc, /grpc-java/compiler/src/test/proto/grpc/testing/compiler/test.proto]
:grpc-compiler:generateTestProto (Thread[Execution worker for ':' Thread 5,5,main]) completed. Took 0.012 secs.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':grpc-compiler:generateTestProto'.
> protoc: stdout: . stderr: /grpc-java/compiler/build/exe/java_plugin/protoc-gen-grpc-java: program not found or is not executable
  Please specify a program using absolute path or make sure the program is available in your PATH system variable
```

file info of `protoc-gen-grpc-java`:
```
root@72d0162ce6ca:/grpc-java# file /grpc-java/compiler/build/exe/java_plugin/protoc-gen-grpc-java
/grpc-java/compiler/build/exe/java_plugin/protoc-gen-grpc-java: ELF 64-bit LSB  executable, ARM aarch64, version 1 (GNU/Linux), dynamically linked (uses shared libs), for GNU/Linux 3.7.0, BuildID[sha1]=ec592fb114f07eff18bf705e0d7bbcd4d042636b, stripped
```

**Comment by surechen** [comment]
> Hi @ejona86,
> I have updated the change #6441, now the script can successfully cross compile the `protoc-gen-grpc-java` artifact, but after bulding `protoc-gen-grpc-java` the building process will use the artifact as plugin of `protoc` to generate the tests proto files, because the compiled `protoc-gen-grpc-java` is for `aarch64` platform, so it cannot be executed in x86_64 server. see the error:
> 
> ```
> > Task :grpc-compiler:generateTestProto FAILED
> Caching disabled for task ':grpc-compiler:generateTestProto' because:
>   Build cache is disabled
> Task ':grpc-compiler:generateTestProto' is not up-to-date because:
>   Task has failed previously.
> [/root/.gradle/caches/modules-2/files-2.1/com.google.protobuf/protoc/3.10.0/e7ba2967f692c3bd2bd417f6a305bcb6a8c6a357/protoc-3.10.0-linux-x86_64.exe, -I/grpc-java/compiler/src/test/proto, -I/grpc-java/compiler/build/extracted-protos/test, -I/grpc-java/compiler/build/extracted-include-protos/test, --java_out=/grpc-java/compiler/build/generated/source/proto/test/java, --plugin=protoc-gen-grpc=/grpc-java/compiler/build/exe/java_plugin/protoc-gen-grpc-java, --grpc_out=/grpc-java/compiler/build/generated/source/proto/test/grpc, /grpc-java/compiler/src/test/proto/grpc/testing/compiler/test.proto]
> :grpc-compiler:generateTestProto (Thread[Execution worker for ':' Thread 5,5,main]) completed. Took 0.012 secs.
> 
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':grpc-compiler:generateTestProto'.
> > protoc: stdout: . stderr: /grpc-java/compiler/build/exe/java_plugin/protoc-gen-grpc-java: program not found or is not executable
>   Please specify a program using absolute path or make sure the program is available in your PATH system variable
> ```
> 
> file info of `protoc-gen-grpc-java`:
> 
> ```
> root@72d0162ce6ca:/grpc-java# file /grpc-java/compiler/build/exe/java_plugin/protoc-gen-grpc-java
> /grpc-java/compiler/build/exe/java_plugin/protoc-gen-grpc-java: ELF 64-bit LSB  executable, ARM aarch64, version 1 (GNU/Linux), dynamically linked (uses shared libs), for GNU/Linux 3.7.0, BuildID[sha1]=ec592fb114f07eff18bf705e0d7bbcd4d042636b, stripped
> ```
> 
> How about skip the testsing steps for `aarch_64` arch ?

I have the same problem. I want to use grpc-java on arm64 platform. Have you solved it?

**Comment by ejona86** [comment]
Fixed by #6441. I'll backport this to v1.26.x branch so it will be in v1.26.0.
